### ISSUE TITLE
Replace orphaned django-ztask with Huey

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-core#2276]
Our current background process scheduler django-ztaskd has been an orphaned project for some time. This in turn hampers our Django and Python update plans, detailed in the following issues and associated comments:
### "Update Django within current Python 2 constrains" #2254

https://github.com/rockstor/rockstor-core/issues/2254#issuecomment-771667824
https://github.com/rockstor/rockstor-core/issues/2254#issuecomment-771823905
https://github.com/rockstor/rockstor-core/issues/2254#issuecomment-771945822

### "move to python 3" #1877
https://github.com/rockstor/rockstor-core/issues/1877#issuecomment-619563667

This issue proposes Huey as a hopefully near fit, well maintained, Python 2.7/3, and Django native without plugins replacement:
https://github.com/coleifer/huey
https://pypi.org/project/huey/
Latest doc links: https://huey.readthedocs.io/en/latest/index.html
v1.10.1 (july 2018) started auto testing on Python 3.7: main readme only mentions 3.4. So good on newer Python.
v2.1.0 (June 2019) "Fix gzip compatibility issue when using Python 2.x."
https://github.com/coleifer/huey/blob/master/CHANGELOG.md#210
So presumably still Python 2.7 at that point at least.

This issue was originally envisaged as a sub part of the above referenced Django update. That way we start out with a far newer version, given the project states supporting/testing only on currently supported versions of Django. So if done before our update we may end up implementing a considerably younger, and possibly more problematic version. With the added work of then possibly having to migrate to newer versions once we update our Django. However we have a strict co-dependence chicken-egg situation here. And as proposed in the original text of the above Django update, we may very well require a testing branch within which we partially complete the Django update with significant breakage, and then follow-up with this issue to complete our ability to achieve even running rockstor services.

**Comment by phillxnet** [comment]
Thanks to forum member freaktechnik in the following forum thread for highlighting a potential current issue with django-ztask's stability:
https://forum.rockstor.com/t/ztaskd-broken-after-leap-updates/7493/2

**Comment by phillxnet** [comment]
Given the prior referenced forum report, I am currently assessing the viability of bringing this issue into our current Rockstor 4 milestone as it  may well be the cause of a recent regression in on our Leap base caused by an inevitable incompatibility in a dependency that was updated upstream. So we may just have to tend to this sooner rather than later for our current future (with existing Django and Python if possible) ready for our 4 Stable release. Rather than our future future (next testing channel). This all depends on if we can find a workable version of Huey that is both compatible with our old Python and Django. If so then we also deal with the django-ztask blocker to updating both Python and Django while addressing this potentially more pressing issue.

**Comment by phillxnet** [comment]
OK folks, we now have somewhat scruffy but working Huey managed rock-on stop, start, install, uninstall mechanism with the Web-UI reacting accordingly. Along with Huey managed event driven logging of these tasks. It's a larger re-work of 'stuff' than I had hoped but Huey is far more capable than our prior now defunct django-ztasks and so the re-work should be worth it. Plus we have not choice but to move on as indicated in the  issue text.

I've yet to test the Rock-on 'update' mechanism however. And this currently, in my 'first draft', is conceived of as running in a single thread/task.
@FroggyFlox What is the most demanding instance/test of a single Rock-on update instance? Rock-nets / adding shares ?

TODO:
- Test rockon config update mechanism.
- Test proposed code rework for Huey re balance/resize pool.
- Work on config restore's transition to Huey and it's associated 'special' Rock-on treatments. May need Huey pipelines for this.

**Comment by FroggyFlox** [comment]
Excellent news!



>  @FroggyFlox What is the most demanding instance/test of a single Rock-on update instance? Rock-nets / adding shares ?

I would say **adding shares** simply because altering rocknets can be done *live* (if the publication state of ports is not changed), meaning that we don't uninstall/reinstall the rock-on. In contrast, *adding shares* (or adding/removing labels) will uninstall the rock-on, and then re-install it with the new settings so it will go through "the whole process".



> Work on config restore's transition to Huey and it's associated 'special' Rock-on treatments. May need Huey pipelines for this.

That one scares me a bit to be honest... In our django-ztask "version", it was running as a task within a task so I really had to do some workaround to have everything running in order. Huey might actually simplify things there! I will try to see what those "Huey pipelines" you speak of are, actually... that sounds interesting.

**Comment by phillxnet** [comment]
@FroggyFlox
Thanks for the pointers. That's exactly what I was after.

Re:
> Huey might actually simplify things there!

Yes, I was hoping so actually. Although my first 'wash' was to do as direct a transition as possible. And if that works then go with that for now. And as we progress we can most-likely improve on what we have had to do previously just to get things working.

References for the Huey pipeline capability, essentially serialising otherwise independently threaded tasks:
https://huey.readthedocs.io/en/latest/guide.html?highlight=pipeline#task-pipelines

There's also a task locking capability which is likely to come in as we discover non parallelisations within our 'methods':
https://huey.readthedocs.io/en/latest/guide.html?highlight=pipeline#locking-tasks
i.e. avoid having more than a single instance of an appropriately decorated function running at any one time. Could be a good initial fit for our config save/restore for example I'm thinking.

**Comment by FroggyFlox** [comment]
> i.e. avoid having more than a single instance of an appropriately decorated function running at any one time. Could be a good initial fit for our config save/restore for example I'm thinking.

That does sound exactly like what we need, indeed... nice!



> Yes, I was hoping so actually. Although my first 'wash' was to do as direct a transition as possible. And if that works then go with that for now. And as we progress we can most-likely improve on what we have had to do previously just to get things working.

I completely agree, especially with regards to rock-ons as we wouldn't lose much time by running some steps "serially" if that helps buy some time to later optimize the process.

Huey does seem like a nice replacement here, and thank you so much for tackling this task! I'm rather short on time lately but I'm looking forward to finding some time to look into it and see how it can help us!

**Comment by phillxnet** [comment]
> * Test rockon config update mechanism.

Bar the occasional disappearance from installed list we now have Huey managed rockon config update.

Tested as working both live, no uninstall/install and non lite updates (label add/ share add).

The 'disappearing' of a mid config update rockon  is very likely an artifact of a race condition as our Huey task updates a model just a bit too late for the automated page refresh. The Rock-on is then considered not-installed. Even though it's state is pending-update. I'm going to leave this for now as it's a small bug for the significant gains of removing django-ztask and in doing so we can, in time, gain more capabilities such as streams2 to manage these more real-time feedback mechanism that we currently simply poll via page refresh.

Moving on to balance/re-size pool.

**Comment by FroggyFlox** [comment]
> I'm going to leave this for now as it's a small bug for the significant gains of removing django-ztask and in doing so we can, in time, gain more capabilities such as streams2 to manage these more real-time feedback mechanism that we currently simply poll via page refresh.

I couldn't agree more; this artifact is likely the result of our sub-optimal current implementation so its improvement deserves its own dedicated work later on. I also am excited by the new possibility we gain such as moving away from polling and into proper real-time feedback. This is especially showing for anything related to rock-ons so it will be a very big improvement once we can!

> capabilities such as streams2

Did you mean Channels2?

**Comment by phillxnet** [comment]
@FroggyFlox 
> Did you mean Channels2?

That's the one. And we need at least Python 3.5 for that also, plus the in-progress Django update (awaiting this issues completion), plus it's follow-up post the python update.
Oh well, bit by bit.

**Comment by phillxnet** [comment]
OK, so a little progress on Huey managed balance commands: We now have successful reporting of a failed regular balance thus (hard coded "btrfs-fake" to throw a fully contextual exception):

![first-huey-balance-fail-report](https://user-images.githubusercontent.com/2521585/108605593-df214200-73ac-11eb-9612-7436b69bd419.png)

We are not at full tick just yet thought but it's a start.

**Comment by phillxnet** [comment]
OK, so we just got another win from this Huey move: our previously dis-functional and remarked out balance "End Time" column is now working as intended, thanks to Huey's task event driven Signals callbacks (see last commit if interested):

![balance-table-end-time-column](https://user-images.githubusercontent.com/2521585/108843769-1113ed80-75d3-11eb-8558-55ae75ac4473.png)

I needed to check the function of the signals here so thought I might as well re-surface this Web-UI element while I was at it. Minor scope creep but I think we can over look this one :).

**Comment by phillxnet** [comment]
It turns our that critical common btrfs balance execution failures such as out of space etc were being hidden inside CommandException() reports in our Huey task errors. Sorted by moving to a full traceback via the last commit. We now have a proper surfacing of the btrfs command execution failure as well as the prior more contrived "btrfs-fake" exception:

![improved-balance-fail-report](https://user-images.githubusercontent.com/2521585/109301533-02764200-7830-11eb-93a3-b3b0290be1ea.png)

In the above id 13 was pre improvement where we only surfaced the TaskException 'error' element. 14 is the same error but with traceback used instead. 15 is our prior contrived failure similarly represented via the now preferred 'traceback' key.

**Comment by phillxnet** [comment]
> We are not at full tick just yet thought but it's a start.

Balance, both internal (disk removal) and external disk add with or without raid level change have now been tested successfully to show (where possible) initial, ongoing, and finished status as expected. Including correctly surfacing contrived and real errors in all these scenarios. Full tick administered appropriately ;) . Moving along to config restore related Huey managed tasks.

**Comment by FroggyFlox** [comment]
Excellent news, @phillxnet!

I just had a quick thought on the following:


> Moving along to config restore related Huey managed tasks.

With regards to the restore of rock-ons (as it is the most complex and potentially most inappropriate), a lot of the process does not need to be run as sub-task (as it currently might?). As `docker` itself does not run in parallel anyway, it would simplify a lot to simply fire the entire thing one after the other and prevent the possibility of starting two restore at the same time. This is especially true given the nature of `docker`, of course, but also the fact that some container parameters might depend on other containers so having one in place before starting actually makes a lot of sense.

Bottom line, I think it's a good idea to streamline a lot of that if it is actually possible and helps!

**Comment by phillxnet** [comment]
@FroggyFlox 
> Bottom line, I think it's a good idea to streamline a lot of that if it is actually possible and helps!

OK, thanks. Yes I was wondering if this may help actually, especially now we have, hopefully, more control over serialisation (Huey pipelines), and inadvertent parallelism (task locking). I've yet to dig into this just yet but it's encouraging that it's the last bit in this Huey move. Oh and I believe we can set timeouts on Huey tasks so that may well help with replacing a similar feature you managed to open-code to await rock-ons install I think it was. Should be back on this later today hopefully.

**Comment by phillxnet** [comment]
OK, so I think I'm now ready for pr submission. Prior to that, to avoid all my messy work in progress/experimentation/save-point commits I'll squash all, remove branch and re-assert/push with the same name so we can hopefully keep the final merge from clouding the commit history.

@FroggyFlox I've just done a few tests and it's looking very much like we are pretty much there, or more accurately 'where we were' on this Huey for django-ztask substitution. And also likely just a tad better behaved :) . Not perfect but it digs us out of a dependency on an orphaned/dead task scheduler and onto a very well maintained one with both old and new django/python compatibility. Pull request to follow shortly hopefully.
